cryptolib/signer.go

```diff
@@ -0,0 +1,70 @@
+/*
+Copyright 2020 Google LLC
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package cryptolib
+
+import "errors"
+
+// Signer contains methods to create a signed Attestation.
+type Signer interface {
+	// CreateAttestation creates an Attestation whose signature is generated by
+	// signing the given payload with the private key. For PGP and PKIX, `payload`
+	// should be the raw payload data. For JWT, `payload` should be a serialized
+	// but unsigned token.
+	CreateAttestation(payload []byte, alg SignatureAlgorithm) (*Attestation, error)
+}
+
+type signer struct {
+	PrivateKey  []byte
+	PublicKeyID string
+}
+
+// NewSigner creates a Signer interface. `privateKey` contains the key data for
+// the private key that signs attestations. `publicKeyID` is the ID of the
+// public key that can verify those signatures.
+func NewSigner(privateKey []byte, publicKeyID string) Signer {
```

Comment:
Will this function be able to decide if it should create a PGP, PKIX, or JWT signer?  It seems like it's missing a hint about this.

Also the format of privateKey should be clearly documented (it's OK to do this as we add implementations, but let's just make sure we don't forget).  For example, in case of PGP, is it
(i) ascii armored, or
(ii) ascii armored and then base64 encoded, or
(iii) something else?

Reply:
We have opted to create separate constructors for each of the key modes. We also added descriptions for the formatting of the private key in the PGP and PKIX case. We selected the same formatting that Binauthz expects of the corresponding public key. We will provide documentation for JWT keys as we near/complete implementation.